### ISSUE TITLE
[FEATURE] Deployment manifests for helping install SPDK dependencies, utilities and libraries

### ISSUE THREAD

**Issue by derekbit** [longhorn/longhorn#5739]
## Describe the solution you'd like

A clear and concise description of what you want to happen

**Comment by derekbit** [comment]
Yeah. I'd update the manifest to making hugepage persistent on each node, so the hugepage is automatically allocated after reboot. QA can retest it.

**Comment by yangchiu** [comment]
@derekbit Is the control plane also needed to install SPDK dependencies even though it's tainted with `NoSchedule` and `NoExecute` so there's no workload can run on this node?

I created a cluster with 1 control plane and 3 worker nodes, and followed the quick start guide to get 3 worker nodes configured for SPDK, and then installed Longhorn.
When I tried to enable SPDK by checking `Enable SPDK Data Engine`, an error shown up:
```
failed to set the setting spdk with invalid value true: 
not enough hugepages-2Mi capacity for node ip-10-0-1-239, requested 1Gi, capacity 0
```
But actually `ip-10-0-1-239` is a control plane with taint `NoSchedule` and `NoExecute`so there's no Longhorn related pods running on it:
```
$ kubectl get nodes
NAME            STATUS   ROLES                  AGE   VERSION
ip-10-0-1-239   Ready    control-plane,master   86m   v1.24.7+k3s1

$ kubectl describe node ip-10-0-1-239
Name:               ip-10-0-1-239
Roles:              control-plane,master
Taints:             node-role.kubernetes.io/master=true:NoExecute
                    node-role.kubernetes.io/master=true:NoSchedule
```
I'm not sure if it's the expected behavior. Thanks!

**Comment by innobead** [comment]
Good finding. @derekbit We should check nodes running instance manager pod instead, right?

cc @shuo-wu

**Comment by yangchiu** [comment]
Compared https://even-pecorino-1bc.notion.site/Quick-Start-for-SPDK-Volumes-a189b3f81e23495d973e3a14b7635667 with https://longhorn.io/docs/1.5.0/spdk/quick-start/. The `Install Extra Kernel Modules` part:
```
apt install -y linux-modules-extra-`uname -r`
```
is missing in https://longhorn.io/docs/1.5.0/spdk/quick-start/.

I'm not sure if it's necessary to manually install `linux-modules-extra-`, but if I didn't do that, the `longhorn-nvme-cli-installation` daemonset will fail on Ubuntu 22.04 because it expects there is `nvme-tcp` but there isn't:
```
$ kubectl apply -f https://raw.githubusercontent.com/longhorn/longhorn/master/deploy/prerequisite/longhorn-nvme-cli-installation.yaml
$ kubectl logs longhorn-nvme-cli-installation-sc2h6 -c nvme-cli-installation
modprobe: FATAL: Module nvme-tcp not found in directory /lib/modules/5.19.0-1026-aws
```

cc @derekbit

**Comment by yangchiu** [comment]
@derekbit After execute `longhorn-spdk-setup`, the `nr_hugepages` is set to `512`: https://github.com/longhorn/longhorn/blob/master/deploy/prerequisite/longhorn-spdk-setup.yaml#L8

But the `environment_check.sh` expects it to be `1024`: https://github.com/longhorn/longhorn/blob/master/scripts/environment_check.sh#L472

Which causes environment check failed:
```
[ERROR] Hugepage size is not enough on node ip-10-0-1-123. Expected: 1024, Actual: 512
[ERROR] Hugepage size is not enough on node ip-10-0-1-123. Expected: 1024, Actual: 512
[ERROR] Hugepage size is not enough on node ip-10-0-1-123. Expected: 1024, Actual: 512
[ERROR] Hugepage size is not enough on node ip-10-0-1-123. Expected: 1024, Actual: 512
```

And the node names of error messages are all pointing to the same node `ip-10-0-1-123`, not sure it just printed the wrong node name or it really only checked one node.

**Comment by derekbit** [comment]
> Good finding. @derekbit We should check nodes running instance manager pod instead, right?
> 
> cc @shuo-wu

Yes.

**Comment by derekbit** [comment]
> I'm not sure if it's necessary to manually install `linux-modules-extra-`, but if I didn't do that, the `longhorn-nvme-cli-installation` and `longhorn-spdk-setup`daemonset will fail on Ubuntu 22.04 because it expects there are `nvme-tcp` and `uio_pci_generic` but there aren't:

https://longhorn.io/docs/1.5.0/spdk/quick-start/ is a general document, so `apt install -y linux-modules-extra-`uname -r`` is not added. Users need to check the kernel modules installation by themselves.

**Comment by derekbit** [comment]
> @derekbit After execute `longhorn-spdk-setup`, the `nr_hugepages` is set to `512`: https://github.com/longhorn/longhorn/blob/master/deploy/prerequisite/longhorn-spdk-setup.yaml#L8
> 
> But the `environment_check.sh` expects it to be `1024`: https://github.com/longhorn/longhorn/blob/master/scripts/environment_check.sh#L472
> 
> Which causes environment check failed:
> 
> ```
> [ERROR] Hugepage size is not enough on node ip-10-0-1-123. Expected: 1024, Actual: 512
> [ERROR] Hugepage size is not enough on node ip-10-0-1-123. Expected: 1024, Actual: 512
> [ERROR] Hugepage size is not enough on node ip-10-0-1-123. Expected: 1024, Actual: 512
> [ERROR] Hugepage size is not enough on node ip-10-0-1-123. Expected: 1024, Actual: 512
> ```
> 
> And the node names of error messages are all pointing to the same node `ip-10-0-1-123`, not sure it just printed the wrong node name or it really only checked one node.

Oh...I forgot to update the environment_check.sh...

**Comment by yangchiu** [comment]
@derekbit Just curious. Is there something needed to be modified in `longhorn-manager`? Like the `hugepages-2Mi` thing.

**Comment by derekbit** [comment]
> @derekbit Just curious. Is there something needed to be modified in `longhorn-manager`? Like the `hugepages-2Mi` thing.

No. Only `spdk` setting.

**Comment by derekbit** [comment]
All issues and feedbacks in this ticket are addressed 
[spdk: update quick-start doc](https://github.com/longhorn/website/pull/724)
[spdk: update expected-nr-hugepages to 512 in environment_check.sh](https://github.com/longhorn/longhorn/pull/6068)
[spdk: validate nodes running instance manager pod rather than all nodes longhorn/longhorn-manager#1952](https://github.com/longhorn/longhorn-manager/pull/1952)

**Comment by yangchiu** [comment]
Verified passed on master-head (longhorn a5041e1). Following steps in https://longhorn.io/docs/1.5.0/spdk/quick-start/, spdk volumes can be created and work without problems.
